Settings stream codec

nvlsianpu opened the pull request:
1)
This patch reworks routines used to store and read the settings data.
Provide stream-style encoding and decoding to/from flash, so the the
API only requires a pointer to binary data, and the settings
implementation takes care of encoding/decoding to/from base64 and
writing/reading to/from flash on the fly. This would eliminate the
need of a separate base64 value buffer on the application-side, thereby
further contributing to the stack footprint reduction.

Above changes allows to remove:
  256-byte value length limitation.
  removing enum settings_type usage so all settings data are treated
  now as a byte array (i.e. what's previously SETTINGS_BYTES)
 
Introduced routine settings_val_read_cb for read and decode the
settings data from storage inside h_set handler implementations.
h_set settings handler now provide persistent value's context
used along with read routine instead of immediately value.

----------------------------------------------------------
Also the user code in higher modules will be aligned then.
For this patch all tests from path teste/subsys/settings/fcb pass.

Any new back-end need to implement two I/O handlers:
```
int (*read_cb)(void *ctx, off_t off, char *buf, size_t *len),
int (*write_cb)(void *ctx, off_t off, char const *buf, size_t len)
size_t (*get_len_cb)(void *ctx)
```
and register it using f. `settings_line_io_init`.
Above handler need only to provide RAW access to the back-end storage.
The Introduced codec take ceare about I/O access alignment.

2)
remove Remove the tgt parameter to the export callback
The tgt parameter is unneeded and only complicates the API.

- [x] stream-style encoding
- [x] fcb-backend
- [x] nffs-backend
- [x] tests align
- [x] make base64 encoding optional
- [x] align the user code in higher modules
  - [x] BT storage
  - [x] BT mesh storage
  - [x] mesh onoff_level_lighting_vnd_app sample

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
subsys/settings: stream-style encoding and decoding to/from storage

This patch reworks routines used to store and read the settings data.
Provide stream-style encoding and decoding to/from flash, so the the
API only requires a pointer to binary data, and the settings
implementation takes care of encoding/decoding to/from base64 and
writing/reading to/from flash on the fly. This would eliminate the
need of a separate base64 value buffer on the application-side, thereby
further contributing to the stack footprint reduction.

Above changes allows to remove:
  256-byte value length limitation.
  removing enum settings_type usage so all settings data are treated
  now as a byte array (i.e. what's previously SETTINGS_BYTES)

Introduced routine settings_val_read_cb for read and decode the
settings data from storage inside h_set handler implementations.
h_set settings handler now provide persistent value's context
used along with read routine instead of immediately value.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu commented:
For sure the code still need some clean-up - so pleas comment on general conception and so one (no coding standard session for now).

nvlsianpu commented:
@jhedberg @Laczen @carlescufi @lemrey Pleas review in yours convenience.

Laczen commented:
Hi @nvlsianpu, sorry I did not respond any faster. I am tied up in other work. This is a great improvement to the settings module.

Maybe there is still an opening for a big simplification of the settings module. As I mentioned before the encoding could be removed. The encoding was imo needed for:
1. Storing binary information as string, when also other information (string, int, ...) was stored as string. If you now only have byte char storage this requirement has been removed.
2. Sending information over serial line (export to serial monitor). As this target has been removed encoding is also not required anymore.
3. Storing binary information in a ascii file with line ends. This is only required for nffs (I am not sure it is really required in this case). This requirement remains.

In fcb you could just store the items as "variable_name=" concatenated with the binary data. As you know the total length of an item in fcb getting the length of the binary data is the difference between the length and the location of the first "=". No need for encoding or streamed encoding. The advantage is of course the reduction in data size and the simplicity.

For nffs it might also be possible to do something like this (I have never used nffs). In the worst case nffs would still require encoding but the settings module would be a lot simpler for fcb.

nvlsianpu commented:
@Laczen Thanks for comments - I received similar comments off-channel from my colleges. I  will differentiate algorithm between FCB and FS. 
Idea of this patch is to satisfy points 1,2,4,6 of #7413 .

>Sending information over serial line (export to serial monitor). As this target has been removed encoding is also not required anymore.

What do you mean, what a serial monitor?

Laczen commented:
Hi @nvlsianpu, what I meant by serial monitor was newt_mgr. The implementation in mynewt has the possibility to export to newt_mgr, this would require encoding to be able to do this.

nvlsianpu commented:
@Laczen Sure - the idea is not to keep things like in the MyNewt it used to be. So zephyr-mcumgr-go-app will suport this (remote provisioning) in the appropriate shape.

jhedberg commented on include/settings/settings.h:
```diff
@@ -160,16 +138,17 @@ int settings_save_one(const char *name, char *var);
  * @return 0 on success, non-zero on failure.
  */
-int settings_set_value(char *name, char *val_str);
+int settings_set_value(char *name, void *val_str, size_t len);
```
This looks very confusing to me. If you say "string" then it normally (at least in C) means a nul-terminated string, in which case a separate length is not needed. Also, if you say "string" the type should be `char *` and not `void *`. Could you clarify what's the idea here?

nvlsianpu replied:
I will change the nam as the value is byte array instead of string:
`int settings_set_value(char *name, void *val, size_t len);`

jhedberg replied:
Ok, that makes things clear. Note that `settings_save_one()` has the same issue.

rakons reviewed (changes requested):
This is a change in good direction. If the whole module is going to be rewriten it may be good moment to start using doxygen tags to describe structures here in correct way.

nvlsianpu committed:
tests: subsys: settings: stream encoding tests

Added tests for recently introduced sterem-style encoding.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
tests: subsys: settings: common tests rework

Introduction of stream-style encoding required tests reworks.
Test for checking especially byte-string encoding is not required
anymore as any value is kept as byte-string.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
tests: subsys: settings: fcb-backend tests rework

Introduction of stream-style encoding required rework
of tests.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
tests: subsys: settings: fs-backend tests rework

Introduction of stream-style encoding required rework
of tests.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
settings: make base64 encoding optional

Thanks to previous commit base64 encoding of the settings
value is not required anymore for NFFS and FCB back-end.

This patch makes encoding an option in case it will be required
in the future.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
tests: subsys: settings: extend nffs back-end testing

Since it is possible to test back-end with base64 and without
test suite should be extended.

This patch introduce tests with and without base64 encoding

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
tests: subsys: settings: extend fcb back-end testing

Since it is possible to test back-end with base64 and without
test suite should be extended.

This patch introduce tests with and without base64 encoding

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
settings: add API for check value size

Such API is convinient for check the persistent storage
value size or whether the value is NULL.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
tests: subsys: settings: tests value get len API

This Patch add tests for API in nffs and fcb back-ends.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
bluetooth: host: align code to reworked settings

The Code need to be align after introduction of stream codec to
setting serialization subsystem.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
bluetooth: common: Add BT hexdump

Added BT alisa for dumping binary data to the logg.
The macro will be useful for BT as reworked settings supports r/w of
raw data - which will be widely leveraged in bluetooth storage alignment
to reworked settings.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
bluetooth: host: mesh: align code to reworked settings

The code need to be align after introduction of stream codec to
setting serialization subsystem.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
sample: host: mesh: onoff_level_lighting align code to reworked settings

The Code need to be align after introduction of stream codec to
setting serialization subsystem.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
tests: subsys: settings: fcb-int tests rework

Introduction of stream-style encoding required rework
of tests.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu commented on subsys/settings/src/settings_line.c:
```diff
@@ -87,3 +93,407 @@ int settings_line_make(char *dst, int dlen, const char *name, const char *value)
+				memcpy(&w_buf[w_size], p_enc, add);
+				enc_len -= add;
+				w_size += add;
+				p_enc += add;
```
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/settings/src/settings_line.c:183:11: error: p_enc may be used uninitialized in this function [-Werror=maybe-uninitialized]

nvlsianpu commented:
/recheck

nvlsianpu commented:
recheck

nvlsianpu commented:
cc @vikrant8051 this PR contains mesh example settings code rework
cc @Qbicz settings implamentation and API was changed, please consider it for your future works, any review welcome.

@jhedberg @Vudentz (at last on of you  both) Please review. 
@Laczen @lemrey @sjanc @carlescufi  Please review in your convenience.

vikrant8051 commented:
@nvlsianpu Thanks for making changes in onoff_level_lighting_vnd_app. It looks now more intuitive.

jhedberg reviewed (changes requested):
Generally I very much like the way the code looks after this. Do you have any figures how much we save in call stack usage? Also, how much testing (if any) have you done for the Bluetooth settings with this PR?

nvlsianpu commented on subsys/settings/Kconfig:
```diff
@@ -9,13 +9,24 @@ menuconfig SETTINGS
+
+# Hidden option to enable encoding value using base64
+config SETTINGS_USE_BASE64
+	bool "encoding value using base64"
```
wrong comment - will correct it

nvlsianpu commented:
> Generally I very much like the way the code looks after this. Do you have any figures how much we save in call stack usage? Also, how much testing (if any) have you done for the Bluetooth settings with this PR?

@jhedberg 

I didn't check how much stack this saves (probably will try when have a chance).
I tested this PR roughly with the bluetooth/peripheral sample, I check  especially bounding and CCC functionality.

If you can perform or recommend me some testing suite pleas do.

nvlsianpu committed:
settings: add delete key-value pair API

So far to deleting av existing key-value pair it was
required to storing NULL value using setting_save_one().
This patch introduce more intuitive API which takes only
the name key string.


Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2603,7 +2606,7 @@ int bt_gatt_clear_ccc(u8_t id, const bt_addr_le_t *addr)
 			     (void *)addr);
 
-	return settings_save_one(key, NULL);
+	return settings_save_one(key, NULL, 0);
```
Added `int settings_delete(const char *name);`

nvlsianpu replied:
I will start using i tomorrow

nvlsianpu commented:
@jhedberg  updated

jhedberg commented on subsys/bluetooth/host/mesh/settings.c:
```diff
@@ -995,7 +958,12 @@ static void clear_rpl(void)
-		settings_save_one(path, NULL);
+		err = settings_delete(path);
+		if (err) {
+			BT_DBG("Failed to store cleared RPL");
```
Should be BT_ERR()

jhedberg commented on subsys/bluetooth/host/mesh/settings.c:
```diff
@@ -957,32 +923,29 @@ void bt_mesh_store_seq(void)
-	settings_save_one(path, str);
+	err = settings_save_one(path, &rpl, sizeof(rpl));
+	if (err) {
+		BT_DBG("Failed to store RPL %s value", path);
```
Should be BT_ERR()

jhedberg commented on subsys/bluetooth/host/mesh/settings.c:
```diff
@@ -927,22 +895,20 @@ void bt_mesh_store_iv(bool only_duration)
-		return;
+	err = settings_save_one("bt/mesh/Seq", &seq, sizeof(seq));
+	if (err) {
+		BT_DBG("Failed to stor Seq value");
```
Should be BT_ERR()

jhedberg commented on subsys/bluetooth/host/mesh/settings.c:
```diff
@@ -897,22 +867,20 @@ void bt_mesh_store_net(void)
+		BT_DBG("Failed to store IV value");
+	} else {
+		BT_DBG("Stored IV value");
+		BT_HEXDUMP_DBG(&iv, sizeof(iv), "raw value");
```
I don't think it's worth doing the he hexdump for the raw value. Previously the code was logging the base64 value so I could compare it to what was restored, however the raw value provides little, well, value, here :)

jhedberg commented on subsys/bluetooth/host/mesh/settings.c:
```diff
@@ -897,22 +867,20 @@ void bt_mesh_store_net(void)
-		return;
+	err = settings_save_one("bt/mesh/IV", &iv, sizeof(iv));
+	if (err) {
+		BT_DBG("Failed to store IV value");
```
Should BT_ERR()

jhedberg commented on subsys/bluetooth/host/mesh/settings.c:
```diff
@@ -858,36 +819,45 @@ static void schedule_store(int flag)
-		return;
+	err = settings_save_one("bt/mesh/Net", &net, sizeof(net));
+	if (err) {
+		BT_DBG("Failed to store Network value");
```
BT_ERR()

jhedberg commented on subsys/bluetooth/host/mesh/settings.c:
```diff
@@ -858,36 +819,45 @@ static void schedule_store(int flag)
+
+	err = settings_delete("bt/mesh/Net");
+	if (err) {
+		BT_DBG("Failed to store clear Network");
```
BT_ERR()

jhedberg commented on subsys/bluetooth/host/mesh/settings.c:
```diff
@@ -858,36 +819,45 @@ static void schedule_store(int flag)
+
+	err = settings_delete("bt/mesh/IV");
+	if (err) {
+		BT_DBG("Failed to store clear IV");
```
BT_ERR()

jhedberg commented on subsys/bluetooth/host/mesh/settings.c:
```diff
@@ -499,102 +473,97 @@ static int hb_pub_set(int argc, char **argv, char *val)
 	}
 
-	if (!val) {
+	if (settings_val_get_len_cb(val_ctx) == 0) {
```
If this is supposed to signify a deletion, isn't it simpler to just use NULL?

nvlsianpu replied:
I was thinking about idea you signal here as well. Unfortunately have not time to implement this return NULL feature. I rather will provide this as other path (which will be compatible backward). I also found API for get  value length usefull.

jhedberg replied:
How do you differentiate between zero-length values and deleted entries? I thought that's why the original API was using NULL for this?

nvlsianpu replied:
It was all the same - NULL meant 0 length and vice versa.

jhedberg replied:
So there has never been a way to indicate that an entry was deleted?

jhedberg replied:
I'm a bit confused now, since AFAIK the settings subsystem has a way of doing garbage collection/compression by eliminating deleted entries when reaching the end of the partition. In such a case you need to be able to tell apart valid but zero-length entries from entries that were deleted (i.e. can be discarded).

jhedberg replied:
Issue clarified in slack: no way to differentiate between deleted and zero-length entries. Should be documented, but not a new issue, so could be fixed in separate PR.

jhedberg commented on subsys/bluetooth/host/mesh/settings.c:
```diff
@@ -1295,18 +1262,18 @@ static void store_pending_mod_pub(struct bt_mesh_model *mod, bool vnd)
-	settings_save_one(path, val);
+	err = settings_save_one(path, val, len);
+	if (err) {
+		BT_DBG("Failed to store %s value", path);
```
BT_ERR()

jhedberg commented on subsys/bluetooth/host/mesh/settings.c:
```diff
@@ -1218,74 +1200,59 @@ static void encode_mod_path(struct bt_mesh_model *mod, bool vnd,
-	settings_save_one(path, val);
+	err = settings_save_one(path, groups, count * sizeof(groups[0]));
+	if (err) {
+		BT_DBG("Failed to store %s value", path);
```
BT_ERR()

jhedberg commented on subsys/bluetooth/host/mesh/settings.c:
```diff
@@ -1218,74 +1200,59 @@ static void encode_mod_path(struct bt_mesh_model *mod, bool vnd,
-	settings_save_one(path, val);
+	err = settings_save_one(path, keys, count * sizeof(keys[0]));
+	if (err) {
+		BT_DBG("Failed to store %s value", path);
```
BT_ERR()

jhedberg commented on subsys/bluetooth/host/mesh/settings.c:
```diff
@@ -1121,40 +1106,37 @@ static void store_net_key(struct bt_mesh_subnet *sub)
-	settings_save_one(path, str);
+	err = settings_save_one(path, &key, sizeof(key));
+	if (err) {
+		BT_DBG("Failed to store AppKey %s value", path);
```
BT_ERR()

jhedberg commented on subsys/bluetooth/host/mesh/settings.c:
```diff
@@ -1121,40 +1106,37 @@ static void store_net_key(struct bt_mesh_subnet *sub)
-	settings_save_one(path, str);
+	err = settings_save_one(path, &key, sizeof(key));
+	if (err) {
+		BT_DBG("Failed to store NetKey value");
```
BT_ERR()

jhedberg commented on subsys/bluetooth/host/mesh/settings.c:
```diff
@@ -1121,40 +1106,37 @@ static void store_net_key(struct bt_mesh_subnet *sub)
+	err = settings_save_one(path, &key, sizeof(key));
+	if (err) {
+		BT_DBG("Failed to store NetKey value");
+
```
Unnecessary empty line

jhedberg commented on subsys/bluetooth/host/mesh/settings.c:
```diff
@@ -1070,48 +1039,64 @@ static void store_pending_cfg(void)
-	settings_save_one(path, NULL);
+	err = settings_delete(path);
+	if (err) {
+		BT_DBG("Failed to store NetKeyIndex 0x%03x",
```
BT_ERR()

jhedberg commented on subsys/bluetooth/host/mesh/settings.c:
```diff
@@ -1070,48 +1039,64 @@ static void store_pending_cfg(void)
-	settings_save_one(path, NULL);
+	err = settings_delete(path);
+	if (err) {
+		BT_DBG("Failed to store AppKeyIndex 0x%03x",
```
BT_ERR(). Does it really not fit on one line?

jhedberg commented on subsys/bluetooth/host/mesh/settings.c:
```diff
@@ -1070,48 +1039,64 @@ static void store_pending_cfg(void)
+
+	err = settings_delete("bt/mesh/Cfg");
+	if (err) {
+		BT_DBG("Failed to store clear configuration");
```
BT_ERR()

jhedberg commented on subsys/bluetooth/host/mesh/settings.c:
```diff
@@ -1070,48 +1039,64 @@ static void store_pending_cfg(void)
-		return;
+	err = settings_save_one("bt/mesh/Cfg", &val, sizeof(val));
+	if (err) {
+		BT_DBG("Failed to store configuration value");
```
BT_ERR()

jhedberg commented on subsys/bluetooth/host/mesh/settings.c:
```diff
@@ -1019,44 +987,45 @@ static void store_pending_rpl(void)
+		BT_DBG("Failed to store Heartbeat Publication");
+	} else {
+		if (to_store) {
+			BT_DBG("Stored Heartbeat Publication");
```
BT_ERR()

jhedberg commented on subsys/bluetooth/host/mesh/settings.c:
```diff
@@ -927,22 +895,20 @@ void bt_mesh_store_iv(bool only_duration)
+		BT_DBG("Failed to stor Seq value");
+	} else {
+		BT_DBG("Stored Seq value");
+		BT_HEXDUMP_DBG(&seq, sizeof(seq), "raw value");
```
Just drop the hexdump, it's not really providing any value

jhedberg commented on subsys/bluetooth/host/mesh/settings.c:
```diff
@@ -1019,44 +987,45 @@ static void store_pending_rpl(void)
+	} else {
+		if (to_store) {
+			BT_DBG("Stored Heartbeat Publication");
+			BT_HEXDUMP_DBG(to_store, sizeof(val), "raw value");
```
Just remove the hexdump

jhedberg commented on subsys/bluetooth/host/mesh/settings.c:
```diff
@@ -159,31 +174,25 @@ static int net_set(int argc, char **argv, char *val)
 
-	BT_DBG("val %s", val ? val : "(null)");
+	if (settings_val_get_len_cb(val_ctx) == 0) {
+		BT_DBG("val (null)");
```
I realize this is just copied from the original log string, but more intuitive would perhaps be "IV deleted" or something like that - isn't that what this branch is about?

jhedberg reviewed (changes requested):
Looks generally good now, thanks! Mostly logging related comments still.

nvlsianpu commented:
@jhedberg I hope I solved all problems you had found.

Laczen reviewed (approved):
Hi @nvlsianpu, very nice work, congrats.

Once this gets merged I will start working on adding a nvs backend. I would like to keep it as simple as possible, but with minimal flash wear:
1. Names are stored in id's ``0x8001...0xBFFF``
2. Corresponding values are stored in id's ``0xC001..0xFFFF``
3. id ``0x8000`` keeps track of the last name id in use.

This way:
a. Flash wear is minimal (only value updates are stored)
b. NVS direct use is possible for id's below 8000.

Settings_NVS would only need to keep track of:
a. settings_nvs_write_id (what id to use next, reusing deleted id's when possible),
b. settings_nvs_last_id (last id to use when reading at start).

carlescufi commented:
@rakons could you please re-review?

rakons reviewed (approved):
I do not like the doxygen style used here to document structure - it might be corrected. But technically it is not an error. Some misssing return value documentation and also the parameters constantness requires thought. I am approving as any of the error here technically is not a something that would make the library unusable.

carlescufi reviewed (approved):
Great job @nvlsianpu
